Mathlib/GroupTheory/GroupAction/Blocks.lean

```diff
@@ -0,0 +1,407 @@
+/-
+Copyright (c) 2024 Antoine Chambert-Loir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir
+-/
+
+import Mathlib.Data.Setoid.Partition
+import Mathlib.GroupTheory.GroupAction.Basic
+import Mathlib.GroupTheory.GroupAction.Pointwise
+import Mathlib.GroupTheory.GroupAction.SubMulAction
+import Mathlib.GroupTheory.Subgroup.Actions
+
+/-! # Blocks
+
+Given `SMul G X`, an action of a type `G` on a type `X`, we define
+
+- the predicate `IsBlock G B` states that `B : Set X` is a block,
+  which means that the sets `g • B`, for `g ∈ G`, are equal or disjoint.
+
+- a bunch of lemmas that give examples of “trivial” blocks : ⊥, ⊤, singletons,
+  and non trivial blocks: orbit of the group, orbit of a normal subgroup…
+
+The non-existence of nontrivial blocks is the definition of primitive actions.
+
+## References
+
+We follow [wieland1964].
+
+-/
+
+open scoped BigOperators Pointwise
+
+namespace MulAction
+
+section orbits
+
+variable {G : Type*} [Group G] {X : Type*} [MulAction G X]
+
+theorem orbit.eq_or_disjoint (a b : X) :
+    orbit G a = orbit G b ∨ Disjoint (orbit G a) (orbit G b) := by
+  apply (em (Disjoint (orbit G a) (orbit G b))).symm.imp _ id
+  simp (config := { contextual := true })
+    only [Set.not_disjoint_iff, ← orbit_eq_iff, forall_exists_index, and_imp, eq_comm, implies_true]
+
+theorem orbit.pairwiseDisjoint :
+    (Set.range fun x : X => orbit G x).PairwiseDisjoint id := by
+  rintro s ⟨x, rfl⟩ t ⟨y, rfl⟩ h
+  contrapose! h
+  exact (orbit.eq_or_disjoint x y).resolve_right h
+
+/-- Orbits of an element form a partition -/
+theorem IsPartition.of_orbits :
+    Setoid.IsPartition (Set.range fun a : X => orbit G a) := by
+  apply orbit.pairwiseDisjoint.isPartition_of_exists_of_ne_empty
+  · intro x
+    exact ⟨_, ⟨x, rfl⟩, mem_orbit_self x⟩
+  · rintro ⟨a, ha : orbit G a = ∅⟩
+    exact (MulAction.orbit_nonempty a).ne_empty ha
+
+end orbits
+
+section SMul
+
+variable (G : Type*) {X : Type*} [SMul G X]
+
+-- Change terminology : is_fully_invariant ?
+/-- A fixed block is a fully invariant subset -/
```

Comment:
Could you please expand the docstrings on `def`s in this file, so that they mention the "context" (multiplicative action of `G` on `X`, etc...).

Reply:
I hope this will be OK.